scripts/clean_obsolete_strings.py

```diff
@@ -0,0 +1,152 @@
+                quarantine_path = line.split('=')[1].strip().strip('"')
+
+    # Make sure paths are set correctly
+    try:
+        l10n_clones_path
+    except NameError:
+        print('l10n_clones_path is not defined in the config file')
+        sys.exit(1)
+    try:
+        quarantine_path
+    except NameError:
+        print('quarantine_path is not defined in the config file')
+        sys.exit(1)
+
+    locales = [x for x in os.listdir(
+        l10n_clones_path) if not x.startswith('.')]
+    # Exclude locales still working on Mercurial directly
+    excluded_locales = [
+        'da', 'it', 'ja', 'ja-JP-mac',
+    ]
+    locales = [l for l in locales if not l in excluded_locales]
+    locales.sort()
+
+    # Get a list of supported files in the source repository
+    source_file_list = extractFileList(quarantine_path)
+
+    for locale in locales:
+        print('Updating: {}'.format(locale))
+        locale_path = os.path.join(l10n_clones_path, locale)
+        target_file_list = extractFileList(locale_path)
+
+        # Update locale repository
+        subprocess.run([
+            'hg', '-R', locale_path, 'pull', '-u'
+        ])
+
+        # Ignore files in locale that are not available in the source
+        target_file_list = [
+            f for f in target_file_list if f in source_file_list]
+
+        # Read source and target and write the output overwriting the existing
+        # localized file
+        for filename in target_file_list:
+            source_filename = os.path.join(quarantine_path, filename)
+            target_filename = os.path.join(locale_path, filename)
+            with open(source_filename) as f:
+                source_content = f.read()
+                source_parser = getParser(filename)
+                source_parser.readUnicode(source_content)
+                reference = list(source_parser.walk())
+            with open(target_filename) as f:
+                target_content = f.read()
+                target_parser = getParser(filename)
+                target_parser.readUnicode(target_content)
+                target = list(target_parser.walk())
+
+            output = serialize(filename, reference, target, {})
+
+            # zh-CN has an extra string that needs to be kept (start page)
+            if locale == 'zh-CN' and filename == 'browser/chrome/browser/browser.properties':
+                output += b'\n# DO NOT REMOVE: this string is used to set up the home page for zh-CN\n'
+                output += b'browser.startup.homepage = https://start.firefoxchina.cn\n'
+
+            with open(target_filename, 'wb') as f:
+                f.write(output)
+
+        # Commit changes
+        subprocess.run([
+            'hg', '-R', locale_path, 'addremove'
+        ])
+        subprocess.run([
+            'hg', '-R', locale_path, 'commit', '-m',
+            'Remove obsolete strings'
+        ])
+        subprocess.run([
+            'hg', '-R', locale_path, 'push'
```

Comment:
I wonder, is it worth to have a `dry` option for this script?

Reply:
It might be useful. Right now I'm editing every time the script, commenting out pieces.

Reply:
Added a bunch of options. I think running dry by default is safer, since it does a lot.